sbi/simulators/mp_simutils.py

```diff
@@ -0,0 +1,293 @@
+    """
+    num_samples = len(theta)
+
+    for i in range(0, num_samples - worker_batch_size + 1, worker_batch_size):
+        yield theta[i : i + worker_batch_size]
+
+    rem_i = num_samples - (num_samples % worker_batch_size)
+    if rem_i != num_samples:
+        yield theta[rem_i:]
+
+
+# todo: do we need seeding back? Or find if we just torch.manual_seed(0)?
+# def reseed(self, seed):
+#     """Carries out the following operations, in order:
+#     1) Reseeds the master RNG for the generator object, using the input seed
+#     2) Reseeds the prior from the master RNG. This may cause additional
+#     distributions to be reseeded using the prior's RNG (e.g. if the prior is
+#     a mixture)
+#     3) Reseeds the simulator RNG, from the master RNG
+#     4) Reseeds the proposal, if present
+#     """
+#     self.rng.seed(seed=seed)
+#     self.seed = seed
+#     self.prior.reseed(self.gen_newseed())
+#     for m in self.models:
+#         m.reseed(self.gen_newseed())
+#     if self.proposal is not None:
+#         self.proposal.reseed(self.gen_newseed())
+
+
+def log(msg, verbose: bool = False):
+    """
+    Print information about the current state of the worker.
+
+    Args:
+        msg: message to print
+        verbose: whether to print the information or not.
+    """
+    if verbose:
+        print("Parent: {}".format(msg))
+
+
+class Worker(mp.Process):
+    def __init__(
+        self,
+        n: int,
+        queue: mp.Queue,
+        pipe: mp.Pipe(),
+        simulator: Callable,
+        seed: int = None,
+        verbose: bool = False,
+    ):
+        """
+        Args:
+            n: worker index
+            queue: queue
+            pipe: pipe
+            simulator: simulator to run on worker
+            seed: seed
+            verbose: whether to give information about the current state of the worker
+        """
+        super().__init__()
+        self.n = n
+        self.queue = queue
+        self.verbose = verbose
+        self.pipe = pipe
+        self.model = simulator
+        self.rng = np.random.RandomState(seed=seed)
+        # TODO seeding not implemented yet: https://github.com/mackelab/sbi/issues/166
+
+    def update(self, i):
+        self.queue.put(i)
+
+    def run(self):
+        """
+        Run simulations on worker.
+
+        Updates simulation results in queue
+        """
+        self.log("Starting worker")
+        while True:
+            try:
+                self.log("Listening")
+                theta_worker_batch = self.pipe.recv()
+            except EOFError:
+                self.log("Leaving")
+                break
+            if len(theta_worker_batch) == 0:
+                self.log("Skipping")
+                self.pipe.send(([], []))
+                continue
+
+            # run forward model for all params, each n_reps times
+            self.log("Received data of size {}".format(len(theta_worker_batch)))
```

Comment:
f-string (did I not say in previous review?)

Reply:
Will change this in a separate commit later when I get rid of `self.log()` in favor of the `logging` module